pkg/model/diamond.go

```diff
@@ -0,0 +1,265 @@
+
+	// SplitFailed is the state of a failed split
+	SplitFailed SplitState = "failed"
+
+	// SplitRunning is the state of running split
+	SplitRunning SplitState = "running"
+)
+
+// IsValid checks the value of a diamond state
+func (s SplitState) IsValid() bool {
+	switch s {
+	case SplitDone, SplitFailed, SplitRunning:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s SplitState) String() string {
+	return string(s)
+}
+
+// DiamondDescriptor models a diamond's metadata
+type DiamondDescriptor struct {
+	DiamondID       string            `json:"diamondID" yaml:"diamondID"`
+	StartTime       time.Time         `json:"startTime" yaml:"startTime"` // when the diamond was initialized
+	EndTime         time.Time         `json:"endTime,omitempty" yaml:"endTime,omitempty"`
+	StateChangeTime time.Time         `json:"stateChangeTime" yaml:"stateChangeTime"` // last state change
+	State           DiamondState      `json:"state" yaml:"state"`
+	Mode            ConflictMode      `json:"mode" yaml:"mode"`
+	HasConflicts    bool              `json:"hasConflicts,omitempty" yaml:"hasConflicts,omitempty"`
+	HasCheckpoints  bool              `json:"hasCheckpoints,omitempty" yaml:"hasCheckpoints,omitempty"`
+	Tag             string            `json:"tag,omitempty" yaml:"tag,omitempty"`
+	BundleID        string            `json:"bundleID,omitempty" yaml:"bundleID,omitempty"`
+	Splits          []SplitDescriptor `json:"splits,omitempty" yaml:"splits,omitempty"`
+	_               struct{}
+}
+
+// DiamondDescriptors is a sortable slice of DiamondDescriptor
+type DiamondDescriptors []DiamondDescriptor
+
+func (b DiamondDescriptors) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+func (b DiamondDescriptors) Len() int {
+	return len(b)
+}
+func (b DiamondDescriptors) Less(i, j int) bool {
+	if !b[i].StartTime.IsZero() && !b[j].StartTime.IsZero() {
+		return b[i].StartTime.Before(b[j].StartTime)
+	}
+	return b[i].DiamondID < b[j].DiamondID
+}
+
+// Last bundle descriptor in slice
+func (b DiamondDescriptors) Last() DiamondDescriptor {
+	return b[len(b)-1]
+}
+
+func defaultDiamondDescriptor() *DiamondDescriptor {
+	ksuid, err := ksuid.NewRandom()
+	if err != nil {
+		panic(fmt.Sprintf("cannot generate random ksuid: %v", err))
+	}
+	return &DiamondDescriptor{
+		DiamondID: ksuid.String(),
+		StartTime: GetBundleTimeStamp(),
+		State:     DiamondInitialized,
+		Mode:      EnableConflicts,
+	}
+}
+
+// NewDiamondDescriptor builds a new DiamondDescriptor
+func NewDiamondDescriptor(opts ...DiamondDescriptorOption) *DiamondDescriptor {
+	d := defaultDiamondDescriptor()
+	for _, apply := range opts {
+		apply(d)
+	}
+	return d
+}
+
+// SplitDescriptor models the metadata about a given split within a diamond
+type SplitDescriptor struct {
+	SplitID               string        `json:"splitID" yaml:"splitID"`
+	StartTime             time.Time     `json:"startTime" yaml:"startTime"`
+	EndTime               time.Time     `json:"endTime,omitempty" yaml:"endTime,omitempty"`
```

Comment:
End time and start time in the same file..

Reply:
Yes and I don't see what's wrong with that, given that the whole thing about diamond is that they are updated (which is, I agree, a novelty when compared to the previous usage of metadata).